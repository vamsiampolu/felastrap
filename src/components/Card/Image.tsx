import { useFela } from "react-fela";
import type { ImageProps, StyleProps, Theme } from "../common";
import { getBorderRadius } from "./imageStyle";

export type CardImageProps = ImageProps &
  Partial<{
    top: boolean;
    bottom: boolean;
  }>;

const base = () => ({ width: "100%" });
const topRule = ({ top }: StyleProps<CardImageProps>) =>
  top ? getBorderRadius({ top }) : {};
const bottomRule = ({ bottom }: StyleProps<CardImageProps>) =>
  bottom ? getBorderRadius({ bottom }) : {};

export const CardImage = ({
  className = "",
  top,
  bottom,
  ...other
}: CardImageProps) => {
  const { css } = useFela<Theme, CardImageProps>({ top, bottom });
  const cx = `${css([base, topRule, bottomRule])} ${className || ""}`.trim();
  return <img data-testid="card-image" className={cx} {...other} />;
};
