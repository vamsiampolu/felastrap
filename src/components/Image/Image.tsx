import { ComponentPropsWithoutRef } from "react";
import { useFela } from "react-fela";
import { StyleProps, Theme } from "../common";

export type CustomProps = Partial<{
  center: boolean;
  figure: boolean;
  fluid: boolean;
  rounded: boolean;
  thumbnail: boolean;
}>;

export type ImageProps = ComponentPropsWithoutRef<"img"> & CustomProps;

const figureRule = ({ figure }: StyleProps<ImageProps>) =>
  figure
    ? {
        marginBottom: "0.5rem",
        lineHeight: 1,
      }
    : {};

const fluidRule = ({ fluid }: StyleProps<ImageProps>) =>
  fluid
    ? {
        maxWidth: "100%",
        height: "auto",
      }
    : {};

const thumbnailRule = ({ thumbnail }: StyleProps<ImageProps>) =>
  thumbnail
    ? {
        padding: "0.25rem",
        backgroundColor: "#fff",
        borderWidth: "1px",
        borderStyle: "solid",
        borderColor: "#dee2e6",
        borderRadius: "0.375rem",
        maxWidth: "100%",
        height: "auto",
      }
    : {};

const centerRule = ({ center }: StyleProps<ImageProps>) => {
  if (center) {
    return {
      display: "block",
      marginRight: "auto",
      marginLeft: "auto",
    };
  }

  return {};
};

const roundedRule = ({ rounded }: StyleProps<ImageProps>) => {
  return rounded ? { borderRadius: "0.375rem" } : {};
};

export const Image = ({
  className = "",
  center = false,
  fluid = false,
  figure = false,
  rounded = false,
  thumbnail = false,
  ...other
}: ImageProps) => {
  const { css } = useFela<Theme, ImageProps>({
    center,
    thumbnail,
    fluid,
    figure,
    rounded,
  });
  const cx = `${css([
    fluidRule,
    thumbnailRule,
    centerRule,
    roundedRule,
    figureRule,
  ])} ${className || ""}`.trim();

  return <img data-testid="image" className={cx} {...other} />;
};
