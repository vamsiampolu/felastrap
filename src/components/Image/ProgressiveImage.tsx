import {
  useRef,
  useEffect,
  type ComponentPropsWithoutRef,
  useState,
} from "react";
import { useFela } from "react-fela";

const placeholderRule = ({ variant = "default" }) =>
  variant === "default"
    ? {
        "0%": {
          backgroundColor: "#eee",
        },
        "50%": {
          backgroundColor: "#ccc",
        },
        "100%": {
          backgroundColor: "#eee",
        },
      }
    : {};

const fadeInRule = ({ variant = "default " }) =>
  variant === "default"
    ? {
        "0%": {
          opacity: 0,
        },
        "100%": {
          opacity: "100%",
        },
      }
    : {};

export type ProgressiveImageProps = Omit<
  ComponentPropsWithoutRef<"img">,
  "loading"
> & { variant?: Variant };

export type Variant = "default" | "blur";
export const ProgressiveImage = ({
  src,
  width,
  height,
  variant = "default",
  ...props
}: ProgressiveImageProps) => {
  const ref = useRef<HTMLImageElement | null>(null);
  const { css, renderer } = useFela();

  const placeholder = renderer.renderKeyframe(placeholderRule, { variant });
  const fadeIn = renderer.renderKeyframe(fadeInRule, { variant });

  const base = css({
    width,
    height,
    "> img": { width: "100%" },
  });

  const loadingAnimation = css({
    animation: `${placeholder} ease-in-out 2s infinite`,
  });

  const fadeInAnimation = `${css({ animation: `${fadeIn} linear 0.5s` })}`;

  const [source, setSource] = useState(src);
  const [className, setClassName] = useState(`${base} ${loadingAnimation}`);

  useEffect(() => {
    if (ref.current?.complete) {
      setClassName(base);
    }

    ref.current?.addEventListener("error", () => {
      setClassName(fadeInAnimation);
      setSource("/error-image.png");
    });
  }, [src, base, fadeInAnimation]);

  return (
    <div data-testid="progressive-image-wrapper" className={className}>
      <img
        data-testid="progressive-image"
        className={fadeInAnimation}
        src={source}
        ref={ref}
        loading="lazy"
        {...props}
      />
    </div>
  );
};
