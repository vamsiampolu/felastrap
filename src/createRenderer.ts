import { createRenderer as createFelaRenderer } from "fela";
import expandShorthand from "fela-plugin-expand-shorthand";
import placeholderPrefixer from "fela-plugin-placeholder-prefixer";
import prefixer from "fela-plugin-prefixer";
import sortClassNames from "fela-sort-classnames";
import sortMediaQueryMobileFirst from "fela-sort-media-query-mobile-first";

export const createRenderer = () => {
  const autoMergeLonghands = true;
  return createFelaRenderer({
    plugins: [
      prefixer(),
      placeholderPrefixer(),
      expandShorthand(autoMergeLonghands),
    ],
    enhancers: [sortClassNames(), sortMediaQueryMobileFirst()],
  });
};
