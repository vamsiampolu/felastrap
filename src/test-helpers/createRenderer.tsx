import expandShorthand from "fela-plugin-expand-shorthand";
import sortClassNames from "fela-sort-classnames";
import sortMediaQueryMobileFirst from "fela-sort-media-query-mobile-first";
import prefixer from "fela-plugin-prefixer";
import placeholderPrefixer from "fela-plugin-placeholder-prefixer";
import { createRenderer as createFelaRenderer, type IRenderer } from "fela";
import { lightTheme } from "../reboot/lightTheme";

export const createRenderer = () => {
  const autoMergeLonghands = true;
  const enhancers = [sortClassNames(), sortMediaQueryMobileFirst()];
  const plugins = [
    prefixer(),
    placeholderPrefixer(),
    expandShorthand(autoMergeLonghands),
  ];

  return createFelaRenderer({
    plugins,
    enhancers,
  });
};

export const cssReboot = (renderer: IRenderer) => {
  renderer.renderStatic(lightTheme);
};
