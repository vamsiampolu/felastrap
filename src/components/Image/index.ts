/* istanbul ignore file */
import { Image, type ImageProps } from "./Image";
import {
  ProgressiveImage,
  type ProgressiveImageProps,
} from "./ProgressiveImage";

export { Image, ProgressiveImage, type ImageProps, type ProgressiveImageProps };
