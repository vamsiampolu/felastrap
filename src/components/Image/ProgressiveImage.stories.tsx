import type { StoryObj, Meta } from "@storybook/react";
import { ProgressiveImage } from "../Image/ProgressiveImage";

export type Story = StoryObj<typeof ProgressiveImage>;

const meta: Meta<typeof ProgressiveImage> = {
  title: "components/ProgressiveImage",
  tags: ["autodocs"],
  component: ProgressiveImage,
};

export default meta;

export const Example: Story = {
  args: {
    width: "200px",
    height: "300px",
    src: "https://picsum.photos/seed/crayon/200/300",
  },
};
