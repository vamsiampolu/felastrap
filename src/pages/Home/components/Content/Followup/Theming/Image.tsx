import { useFela } from "react-fela";
import { Column } from "../../../../../../components/Flex";
import { Image as BaseImage } from "../../../../../../components/Image/Image";
import themes from "../../../../../../assets/bootstrap-themes.png";
import themes2x from "../../../../../../assets/bootstrap-themes@2x.png";

export const Image = () => {
  const { css } = useFela();
  return (
    <Column lg={6}>
      <BaseImage
        className={`${css({
          marginTop: "1rem",
          marginRight: "auto",
          marginLeft: "auto",
        })} bootstrap-themes-image`}
        fluid
        loading="lazy"
        alt="Bootstrap Themes"
        srcSet={`${themes}, ${themes2x}`}
        src={themes}
        width={700}
        height={500}
      />
    </Column>
  );
};
