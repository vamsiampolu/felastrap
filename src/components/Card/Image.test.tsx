import { IRenderer } from "fela";
import { CardImage } from "./Image";
import { createRenderer, cssReboot } from "../../test-helpers/createRenderer";
import { createSnapshot } from "../../test-helpers/createSnapshot";

describe("components/Card/Image", () => {
  let renderer: IRenderer;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });

  it("should render an image at the top of the card", async () => {
    const component = <CardImage top src="https://placehold.co/600x400" />;
    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render an image at the bottom of the card", async () => {
    const component = <CardImage bottom src="https://placehold.co/600x400" />;
    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render an image as part of the card body", async () => {
    const component = (
      <CardImage
        bottom={false}
        top={false}
        src="https://placehold.co/600x400"
      />
    );
    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });
});
