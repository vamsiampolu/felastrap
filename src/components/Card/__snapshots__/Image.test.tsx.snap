// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Card/Image should render an image as part of the card body 1`] = `
".a {
  width: 100%;
}


<img data-testid="card-image" class="a" src="https://placehold.co/600x400" />;
"
`;

exports[`components/Card/Image should render an image at the bottom of the card 1`] = `
".a {
  width: 100%;
}
.b {
  border-bottom-right-radius: calc(0.375rem - 1px);
}
.c {
  border-bottom-left-radius: calc(0.375rem - 1px);
}


<img
  data-testid="card-image"
  class="a b c"
  src="https://placehold.co/600x400"
/>;
"
`;

exports[`components/Card/Image should render an image at the top of the card 1`] = `
".a {
  width: 100%;
}
.b {
  border-top-left-radius: calc(0.375rem - 1px);
}
.c {
  border-top-right-radius: calc(0.375rem - 1px);
}


<img
  data-testid="card-image"
  class="a b c"
  src="https://placehold.co/600x400"
/>;
"
`;
